### ISSUE TITLE
Relation Classiﬁcation via Convolutional Deep Neural Network (COLING-2014)

### ISSUE THREAD

**Issue by BrambleXu** [BrambleXu/knowledge-graph-learning#13]
# 一句话总结

第一个使用了word embedding，以及positional mebdding，并把CNN用于RE任务上的工作。

# 资源

- [code by PyTorch](https://github.com/ShomyLiu/pytorch-pcnn)

# 概要

## 要解决的问题

此前的RE模型都是基于统计模型的，效果的好坏很大程度上取决于抽取的特征。如果特征抽取不好的话，误差也会被传递下去。

如何提取好的特征是一个问题？

## 解决方法

通过DNN来lexical和sentence level的特征。

- 不对word tokens进行任何预处理。
- 通过embedding将word变为vector
- lexical level features are extracted according to the given nouns
- 通过CNN抽取sentence level feature
- 把这两种feature结合起来，作为最终抽取出来的特征向量
- 把特征向量喂给softmax分类器，预测两个标记的名字之间的关系

## 一些注意点

- RE是作为一个多分类问题来被处理的。
- 因为是找relation的任务，所以position feature(PF) of noun pairs是需要利用起来的。

# 3 Methodology

## 3.1 The Neural Network Architecture

![image-20190221160047337](https://ws2.sinaimg.cn/large/006tKfTcgy1g0e2161j70j30cl0a3dhk.jpg)

主要有三个部分：Word Representation, Feature Extraction and Output.

## 3.2 Word Representation

> Our experiments directly utilize the trained embeddings provided by Turian et al.(2010).

这个embedding找不到了

## 3.3 Lexical Level Features

We select the word embeddings of marked nouns and the context tokens. Moreover, the WordNet hypernyms 4 are adopted as MVRNN (Socher et al., 2012). All of these features are concatenated into our lexical level features vector

![image-20190221161222351](https://ws3.sinaimg.cn/large/006tKfTcgy1g0e2d6ya5pj30a7058t99.jpg)

## 3.4 Sentence Level Features

用max-pooled CNN抽取特征。

![image-20190221161557434](https://ws4.sinaimg.cn/large/006tKfTcgy1g0e2gxdw40j30i20er41v.jpg)

- In the Window Processing component, each token is further represented as Word Features (WF) and Position Features (PF) (see section 3.4.1 and 3.4.2). 
- Then, the vector goes through a convolutional component. 
- Finally, we obtain the sentence level features through a non-linear transformation.

### 3.4.1 Word Features

![image-20190221161919297](https://ws2.sinaimg.cn/large/006tKfTcgy1g0e2kfccagj30oj03bq4c.jpg)

![image-20190221161948226](https://ws2.sinaimg.cn/large/006tKfTcgy1g0e2kygvyzj30oo040jsu.jpg)

一对entity对应一个label y。window size是3.

### 3.4.2 Position Features

structure features (e.g., the shortest dependency path between nominals) 是一个很有用的feautre，可以包含两个entity之间的依赖信息。但是这种特征无法通过上面的CNN提取到，所以PF被提出来了。

![image-20190221162629613](https://ws1.sinaimg.cn/large/006tKfTcgy1g0e2rw52lvj30oc05776u.jpg)

比如上面的例子，moving到People和downtown的距离是3, -3。而且这个相对距离会被映射到一个$d_e$维的向量。于是得到$\rm PF=[\mathbf{d}_1, \mathbf{d}_2] $. 

### 3.4.3 Convolution

![image-20190221164532100](https://ws4.sinaimg.cn/large/006tKfTcgy1g0e3bpoyjzj30or06tju3.jpg)

W1是权重，X是window process输出的结果。

![image-20190221164703141](https://ws2.sinaimg.cn/large/006tKfTcgy1g0e3da8keij30oh02vq3s.jpg)

最后从每个句子抽取，得到一个n1维的向量（代表关系），这样就和句子长度无关了。

### 3.4.4 Sentence Level Feature Vector

为了学习更复杂的特征，这里用了tanh作为非线性激活函数。

![image-20190221164959414](https://ws1.sinaimg.cn/large/006tKfTcgy1g0e3gc4mhxj30ow04iq3u.jpg)

![image-20190221165011740](https://ws1.sinaimg.cn/large/006tKfTcgy1g0e3gjozmfj30oj02s3ze.jpg)

这里说经过激活函数后得到的结果代表更高维的特征，比如代表整个句子

## 3.5 Output

![image-20190221165249963](https://ws4.sinaimg.cn/large/006tKfTcgy1g0e3jar2gcj30oc071ju4.jpg)

## 3.6 Backpropagation Training

![image-20190221165522716](https://ws2.sinaimg.cn/large/006tKfTcgy1g0e3ly4yymj30or05ztao.jpg)

输入是句子s，经过window processing之后，得到了X。N是word embedding. 输出是每个relation的概率。

![image-20190221165809842](https://ws1.sinaimg.cn/large/006tKfTcgy1g0e3ouhh5ej30oy03j752.jpg)

我们的目的是最大化这个目标函数。

![image-20190221165918671](https://ws3.sinaimg.cn/large/006tKfTcgy1g0e3q145exj30ow06g0va.jpg)

优化方法使用SGD。N，W，X都是随机初始化的，X是通过embedding得到的。

# 4 Dataset and Evaluation Metrics

SemEval-2010 Task 8 dataset

macro-averaged F1-scores

# 5 Experiments

## 5.1 Parameter Settings

![image-20190221170529328](https://ws4.sinaimg.cn/large/006tKfTcgy1g0e3wgwn0ij312j0cjwg3.jpg)

![image-20190221170539615](https://ws3.sinaimg.cn/large/006tKfTcgy1g0e3wnbn2mj30ox03ddgn.jpg)

## 5.2 Results of Comparison Experiments

![image-20190221170728895](https://ws2.sinaimg.cn/large/006tKfTcgy1g0e3yjh0yuj30no08u0vp.jpg)

## 5.3 The Effect of Learned Features

![image-20190221170842323](https://ws3.sinaimg.cn/large/006tKfTcgy1g0e3zt5ql7j30ar060aai.jpg)

# 6 Conclusion

position features (PF) are successfully proposed to specify the pairs of nominals to which we expect to assign relation labels. The system obtains a signiﬁcant improvement when PF are added.
